single_file_stac_api/backend.py

```diff
@@ -52,24 +76,84 @@ def insert_collection(self, collection: Collection):
         self.collections.append(collection)
 
 
-@dataclass
-class SingleFileClient(BaseTransactionsClient, BaseCoreClient):
+class PaginationToken(BaseModel):
+    """Pagination model."""
+
+    id: str
+    keyset: str
+
+
+@attr.s
+class PaginationTokenClient:
+    """Pagination token."""
+
+    token_table: List[PaginationToken] = attr.ib(factory=list)
+
+    def insert_token(self, keyset: str, tries: int = 0) -> str:  # type:ignore
+        """Insert a keyset into the database."""
+        # uid has collision chance of 1e-7 percent
+        uid = urlsafe_b64encode(os.urandom(6)).decode()
+        self.token_table.append(PaginationToken(id=uid, keyset=keyset))
+        return uid
+
+    def get_token(self, token_id: str) -> int:
+        """Retrieve a keyset from the database."""
+        rows = list(filter(lambda x: x.id == token_id, self.token_table))
+        if not rows:
+            raise NotFoundError(f"{token_id} not found")
+        return int(rows[0].keyset)
+
+
+@attr.s
+class Paging:
+    """Simple list Paging."""
+
+    items: List = attr.ib()
+    limit: int = attr.ib(default=10)
+
+    def __attrs_post_init__(self):
+        """Post Init."""
+        num_pages = len(list(range(0, len(self.items), self.limit)))
+        self.pages = [
+            Page(
+                items=self.items[i : i + self.limit],
+                num=idx,
+                has_next=(0 <= idx < num_pages - 1),
+                has_previous=(0 < idx),
+            )
+            for idx, i in enumerate(range(0, len(self.items), self.limit))
+        ]
+
+    def get_page(self, page_number=None):
+        """return page."""
+        if not page_number:
+            page_number = 0
+        return list(filter(lambda x: x.num == page_number, self.pages))[0]
+
+
+@attr.s
+class Page:
+    """Simple Page model."""
+
+    items: List = attr.ib()
+    num: int = attr.ib(default=0)
+    has_next: bool = attr.ib(default=False)
+    has_previous: bool = attr.ib(default=False)
+
+
+@attr.s
+class SingleFileClient(BaseTransactionsClient, BaseCoreClient, PaginationTokenClient):
     """application logic"""
 
-    db: Database
+    db: Database = attr.ib(factory=Database)
```

Comment:
if let without default, attrs will complain about non-default parameter being placed after defaults args

Reply:
or you can set the class to keyword only